host/xtest/stats.c

```diff
@@ -209,6 +218,107 @@ static int stat_memleak(int argc, char *argv[])
 	return close_sess(&ctx, &sess);
 }
 
+static int stat_loaded_ta(int argc, char *argv[])
+{
+	TEEC_Context ctx = { };
+	TEEC_Session sess = { };
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	uint32_t eo = 0;
+	TEEC_Operation op = { };
+	void *buff = NULL;
+	struct ta_dump_stats *stats = NULL;
+	size_t stats_size_bytes = 0;
+	size_t n = 0;
+	uint32_t retry_count = 10;
+
+	UNUSED(argv);
+	if (argc != 1)
+		return usage();
+
+	open_sess(&ctx, &sess);
+retry:
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_OUTPUT,
+					 TEEC_NONE, TEEC_NONE, TEEC_NONE);
+	res = TEEC_InvokeCommand(&sess, STATS_CMD_TA_STATS, &op, &eo);
+	if (res != TEEC_ERROR_SHORT_BUFFER) {
+		printf("TEEC_InvokeCommand: res %#"PRIx32" err_orig %#"PRIx32"\n",
+		       res, eo);
+		goto out;
```

Comment:
For all the error case, I think you should report the error. Can use `errx()` as done in other `stat_xxx()` function in this source file. Here:
```
		errx(EXIT_FAILURE,
		     "STATS_CMD_TA_STATS: expected %#"PRIx32", got %#"PRIx32" orig",
		     stats_size_bytes, op.params[1].tmpref.size);

Reply:
`errx()` will call `exit (1);`. But here we might need to `goto out;` and release resource before exit xtest.
I realized that  `errx()` will output log to `stderr`. It seems to me there is no much difference we output to  `stderr` or `stdout`. Like to know which device you would suggest?

Reply:
`stderr` is normally unbuffered, while `stdout` might be buffered. Then it's also different file descriptors so if you redirect the output you'll get different results.

I'd prefer using `errx()` too, I believe all the resources are reclaimed when the process exits any way.

Reply:
thanks, have changed to errx().
like to consult one thing, how the session can be reclaimed when the process exits? When a client open a session, there is corresponding session context -- `struct tee_ta_session` in OP-TEE core nad the session will be put into a link -- `open_sessions.`  When the client exits without calling closeSession(), the session context in OP-TEE core will be reclaimed?

Reply:
> When the client exits without calling closeSession(), the session context in OP-TEE core will be reclaimed?

Yes, the context file descriptor is closed `optee_release()` in the kernel driver will be called. This function tries to close all remaining open sessions belonging to this context.

Reply:
okay, great to know that.